pyne/openmc_utils.py

```diff
@@ -156,3 +171,274 @@ def xml(self):
+
+    Parameters:
+    -----------
+    filename : str
+        The OpenMC state point file name.
+    m : MeshTally
+        MeshTally for the tally.
+
+    Returns:
+    --------
+    result : numpy array
+        This numpy array contains the flux data read from OpenMC state point
+        file. The shape of this numpy array is (num_ves*num_e_groups), with z
+        changes fastest.
+    rel_error: numpy array
+        This numpy array contains the relative error data.
+    res_tot : list
+        The total results.
+    rel_err_tot : list
+        Relative error of total results.
+    """
+
+    sp = openmc.StatePoint(filename)
+    tally = sp.get_tally(scores=['flux'], id=m.tally_number)
+    flux = tally.get_slice(scores=['flux'])
+
+    num_ves = len(m)
+    # currently, the openmc mesh are uniform
+    ve_vol = m.structured_hex_volume(0, 0, 0)
+    num_e_groups = len(flux.mean.flatten()) // num_ves
+
+    # get result and res_tot
+    result = flux.mean.flatten()
+    result = np.reshape(result, newshape=(num_e_groups, num_ves))
+    result = result.transpose()
+    res_tot = np.sum(result, axis=1)
+
+    # calculate rel_err
+    rel_err = np.zeros_like(flux.std_dev)
+    nonzero = flux.mean > 0
+    rel_err[nonzero] = flux.std_dev[nonzero] / flux.mean[nonzero]
+    rel_err = np.reshape(rel_err.flatten(), newshape=(num_e_groups, num_ves))
+    rel_err = rel_err.transpose()
+    # calculate rel_err_tot, assuming independent in each energy bin (not true)
+    # due to the lack of covariance, this could be smaller than the true value
+    rel_err_tot = np.zeros_like(res_tot)
+    std_dev = np.reshape(flux.std_dev.flatten(),
+                         newshape=(num_e_groups, num_ves))
+    std_dev = std_dev.transpose()
+    var_tot = np.sum(np.square(std_dev), axis=1)
+    nonzero = res_tot > 0
+    rel_err_tot[nonzero] = np.sqrt(var_tot[nonzero]) / res_tot[nonzero]
+
+    # get volume averged results
+    result = np.divide(result, ve_vol)
+    res_tot = np.divide(res_tot, ve_vol)
+    # In the result of OpenMC, x changes fastest
+    # change the order to z changes fastest
+    result = result_changes_order(result, m.dims[3:6])
+    rel_err = result_changes_order(rel_err, m.dims[3:6])
+    res_tot = result_changes_order(res_tot, m.dims[3:6])
+    rel_err_tot = result_changes_order(rel_err_tot, m.dims[3:6])
+    return result, rel_err, res_tot, rel_err_tot
+
+def result_changes_order(result, dims):
+    """
+    This function changes the order of openmc flux data. The default order of
+    OpenMC flux is x changes the fastest. This function changes it to the order
+    which z changes the fasest (the same order as MCNP meshtal).
+
+    Parameters:
+    -----------
+    result : numpy array
+        This numpy array contains the flux data read from OpenMC state point
+        file. The shape of this numpy array is (num_ves*num_e_groups) or
+        (num_ves, ), with x changes the fastest.
+    dims : list
```

Comment:
tuple?

Reply:
I found that in class `Mesh`, the `dims` and boundaries are stored as `list` or `list of list`. It's better to make a separate PR to change them to `tuple` at once.

Reply:
Sounds good to me.